lib/fluent/plugin/buffer.rb

```diff
@@ -761,6 +771,12 @@ def write_step_by_step(metadata, data, format, splits_count, &block)
             begin
               while writing_splits_index < splits.size
                 split = splits[writing_splits_index]
+                if split.size == 1 && original_bytesize == 0 && (chunk.bytesize + split.first.to_msgpack.bytesize) > @chunk_limit_size
+                  # When it will be predictable that BufferChunkOverflowError occur,
+                  # commit current chunk to keep it, otherwise valid 'split' data can be lost by
+                  # the following chunk.rollback
+                  chunk.commit
+                end
```

Comment:
This is the point I'm concerned.

* Doesn't it have performance regression?
* It seems that it cares only the case on msgpack.
  * It might not exceed when `compress gzip` case

Reply:
>       * It might not exceed when `compress gzip` case

In my understanding it's harmless because it doesn't exceed the limit as a result.

Reply:
> Doesn't it have performance regression?

Yes, noted in a commit message, it has performance regression.
I'll reconsider another approach again.